src/olympia/lib/git.py

```diff
@@ -0,0 +1,168 @@
+            self.git_repository.create_commit(
+                'HEAD',  # ref
+                self.get_author(),  # author
+                self.get_author(),  # commitor
+                'Initializing repository',  # message
+                tree,  # tree
+                [])  # parents
+
+            log.debug('Initialized git repository {path}'.format(
+                path=self.git_repository_path))
+        else:
+            self.git_repository = pygit2.Repository(self.git_repository_path)
+
+    @classmethod
+    def extract_and_commit_from_file_obj(cls, file_obj, channel):
+        """Extract all files from `file_obj` and comit them.
+
+        This is doing the following:
+
+        * Create a temporary `git worktree`_
+        * Remove all files in that worktree
+        * Extract the zip behind `file_obj` into the worktree
+        * Commit all files
+
+        Kinda like doing...
+
+        * rm -rf worktree/*
+        * unzip file.zip -d worktree/
+        * git commit -a worktree/*
+
+        This ignores the fact that there may be a race-condition of two
+        versions being created at the same time. Since all relevant file based
+        work is done in a temporary worktree there won't be any conflicts and
+        usually the last upload simply wins the race and we're setting the
+        HEAD of the branch (listed/unlisted) to that specific commit.
+
+        .. _`git worktree`: https://git-scm.com/docs/git-worktree
+        """
+        addon = file_obj.version.addon
+        repo = cls(addon.id)
+
+        with TemporaryWorktree(repo.git_repository) as worktree:
+            # Now extract the zip to the workdir
+            zip_file = SafeZip(file_obj.current_file_path, force_fsync=True)
+            zip_file.extract_to_dest(worktree.path)
+
+            # Stage changes
+            worktree.repo.index.add_all()
+            worktree.repo.index.write()
+            tree = worktree.repo.index.write_tree()
+
+            # Create an orphaned commit
+            message = (
+                'Create new version {version} ({version_id}) for '
+                '{addon} from {file_obj}'.format(
+                    version=repr(file_obj.version),
+                    version_id=file_obj.version.id,
+                    addon=repr(addon),
+                    file_obj=repr(file_obj)))
+
+            oid = worktree.repo.create_commit(
+                None,
+                repo.get_author(), repo.get_author(),
+                message,
+                tree,
+                []
+            )
+
+            commit = repo.git_repository.get(oid)
+
+        branch = repo.find_or_create_branch(BRANCHES[channel])
+        branch.set_target(commit.hex)
+
+        # Set the latest git hash on the related version.
+        file_obj.version.update(git_hash=commit.hex)
+        return repo
+
+    def get_author(self):
+        return pygit2.Signature(
+            name='Mozilla Add-ons Robot',
```

Comment:
It's always this exact author?  Seems we should cache this value if it's used multiple times.

Reply:
_Could_ we use the name and email of the developer uploading the xpi?  Right now it won't be used, but I could imagine it might be useful information at some future point to know who uploaded what in a diff.  We might want to expose it.

Reply:
We can only make sure to do that for future uploads since we don't save the information for old uploads. We could correlate that with creation dates of `AddonUser` objects but I'm not sure this will be reliable.

I'm happy to implement that though for future uploads once I'm adding the extraction to the actual upload. For mass-extraction we don't have that information unfortunately.